src/main/java/net/datafaker/service/RandomService.java

```diff
@@ -141,6 +145,84 @@ public Random getRandomInternal() {
+     *              - value: The actual element of type T to be randomly selected based on its weight.
+     * @param <T> The type of the element to be selected from the list. The value associated with the weight can be of any type.
+     * @return A randomly selected element based on its weight.
+     * @throws IllegalArgumentException if the list is null, empty, if any item in the list is null or empty,
+     *                                  if any weight is null, non-positive, NaN, negative zero, infinite,
+     *                                  or if the item does not contain 'weight' or 'value' keys.
+     */
+    public <T> T weightedArrayElement(List<Map<String, Object>> items) {
+        validateItemsList(items);
+
+        double totalWeight = calculateTotalWeight(items);
+        double randomValue = random.nextDouble() * totalWeight;
+
+        return selectWeightedElement(items, randomValue);
+    }
+
+    private void validateItemsList(List<Map<String, Object>> items) {
+        if (items == null || items.isEmpty()) {
+            throw new IllegalArgumentException("weightedArrayElement expects a non-empty list");
+        }
+        items.forEach(this::validateItem);
+    }
+
+    private void validateItem(Map<String, Object> item) {
+        if (item == null || item.isEmpty()) {
+            throw new IllegalArgumentException("Item cannot be null or empty");
+        }
+        if (!item.containsKey(WEIGHT_KEY) || !item.containsKey(VALUE_KEY)) {
+            throw new IllegalArgumentException("Each item must contain 'weight' and 'value' keys");
+        }
+        validateValue(item.get(VALUE_KEY));
+        validateWeight(item.get(WEIGHT_KEY));
+    }
+
+    private void validateValue(Object valueObj) {
+        if (valueObj == null) {
+            throw new IllegalArgumentException("Value cannot be null");
+        }
+    }
+
+    private void validateWeight(Object weightObj) {
+        if (!(weightObj instanceof Double)) {
+            throw new IllegalArgumentException("Weight must be a non-null Double");
+        }
+
+        double weight = (Double) weightObj;
+
+        if (weight <= 0 || Double.isNaN(weight) || Double.isInfinite(weight)) {
+            throw new IllegalArgumentException("Weight must be a positive number and cannot be NaN or infinite");
+        }
+    }
+
+    private double calculateTotalWeight(List<Map<String, Object>> items) {
+        return items.stream()
+            .mapToDouble(item -> (Double) item.get(WEIGHT_KEY))
+            .sum();
+    }
+
+    private <T> T selectWeightedElement(List<Map<String, Object>> items, double randomValue) {
+        double currentWeight = 0.0;
+
+        for (Map<String, Object> item : items) {
+            currentWeight += (Double) item.get(WEIGHT_KEY);
+            if (randomValue < currentWeight) {
+                return (T) item.get(VALUE_KEY);
+            }
+        }
```

Comment:
is there a reason to do this on each iteration?
IIUC data in file is persistent then we could sort them once including all unboxing things (not sure if need them...) only once and have something like array of doubles
then just use something like adopted binary search

or did I miss anything here?

Reply:
Thank you for the suggestion. I implemented the proposed optimization of sorting and preprocessing the cumulative weights once, which improved the efficiency.